Move backward and set_data off of Type

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21964 Remove Type dispatch
* **#21963 Stop using Type in Python bindings**
* #21941 Remove many usages of Type
* #21940 Remove getDeviceFromPtr and allocator from Type
* #21923 Move UnsafeTensorFromTH and UnsafeStorageFromTH off Type
* #21320 [BC-BREAKING] Replace Type dispatch with ATenDispatch

These functions only work on variables; calling it on a Tensor errors out. I moved the Tensor error stubs to native functions, and changed the variable implementations to be manual overrides in VariableType.

Differential Revision: [D15897423](https://our.internmc.facebook.com/intern/diff/D15897423)

Commit:
Stop using Type in Python bindings

Commit:
Update on "Stop using Type in Python bindings"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Stop using Type in Python bindings"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

ezyang reviewed (approved):
The signature change to `backward` seems harmless enough since it was defaulted, but I would like to understand a little more. Everything else looks fine.

gchanan commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -354,19 +354,13 @@ class CAFFE2_API Tensor {
 
   //example
   //Tensor * add(Tensor & b);
+  void backward(const Tensor & gradient={}, bool keep_graph=false, bool create_graph=false) const;
```
we've been planning to make `Tensor?` translate to `optional<Tensor>` for awhile now -- maybe now is a good time to do it?

gchanan commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -68,22 +68,23 @@ void Variable::detach_() {
   std::vector<Variable> inputs;
-  if (!gradient.has_value()) {
-    gradient = at::ones_like(*this);
+  Tensor gradient_;
```
wait, where do you assign `gradient` to `gradient_`?

gchanan commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -252,16 +252,7 @@ int THPVariable_set_grad(THPVariable *self, PyObject *py_grad)
-  }
-
-  THPUtils_assertRet(-1, grad.type() == var.type() || gradIsSparse,
+  THPUtils_assertRet(-1, grad.type() == var.type() || grad.is_sparse(),
```
unless I'm missing something, this doesn't look equivalent?  The new code allows you to assign a sparse gradient, even if it's the wrong dtype.

gchanan commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -191,14 +84,14 @@ std::vector<at::Tensor> VariableType::unpack(at::TensorList tl, const char *name
+    const Tensor& gradient,
     bool keep_graph,
-    bool create_graph) const {
+    bool create_graph) {
```
is there some reason the const disappeared on this `backward` and `set_data`?

gchanan commented on aten/src/ATen/core/TensorMethods.h:
```diff
@@ -56,18 +56,15 @@ inline TensorOptions Tensor::options() const {
+// all static inline to allow for inlining of the non-dynamic part of dispatch
+inline void Tensor::backward(const Tensor & gradient, bool keep_graph, bool create_graph) const {
+    static auto table = globalATenDispatch().getOpTable("aten::backward(Tensor(a!) self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> void");
+    return table->getOp<void (const Tensor &, const Tensor &, bool, bool)>(tensorTypeIdToBackend(type_id()), is_variable())(*this, gradient, keep_graph, create_graph);
```
cool.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -29,6 +29,12 @@
 - func: _cast_Half(Tensor self, bool non_blocking=False) -> Tensor
   variants: function
 
+- func: backward(Tensor(a!) self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> void
```
how did you choose these annotations?  For example, why is backward's `self` marked as writeable, but set_data isn't?

gchanan replied:
I'm not even really sure what these annotations mean in this context, as they are pretty different than the usual meanings.

For example, `backward` is the only place where we mutate `self.grad` -- does that count as mutable?  `self` itself doesn't actually change.

Maybe @suo can give some guidance here.

suo replied:
`backward()` is a special case that isn't expressible in the alias annotation language—it writes to every tensor in the autograd graph as the gradients are accumulated to the leaves. So if we enabled in the JIT it would have to be a special case in the alias analysis pass.

The bit about mutating the `grad` property only isn't that relevant—for the purpose of optimization, you can't reorder a call to `backward()` across a read of any tensor in the autograd graph anyway, so it's fine to consider `self.grad` part of the `self`'s memory.

gchanan reviewed (changes requested):
see comments.

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

gchanan commented on aten/src/ATen/native/VariableMethodStubs.cpp:
```diff
@@ -0,0 +1,16 @@
+  AT_ERROR("backward is not implemented for Tensor");
+}
+
+void set_data(const Tensor& self, const Tensor& new_data) {
```
the codegen for this looks wrong, because this is really an inplace method, so you'd think `self` would be `Tensor &`, not `const Tensor &`, probably because this doesn't end in a _.  I can't off the top of my head think why this would cause a problem, just something to note.

gchanan reviewed (approved):
I don't think the remaining issues around annotations really matter, but I'd still be interested in what the actual correct annotations are.

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

Commit:
Update on "Move backward and set_data off of Type"

Move backward and set_data off of Type

gh-metadata: pytorch pytorch 21963 gh/li-roy/31/head

soumyarooproy commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -252,15 +252,7 @@ int THPVariable_set_grad(THPVariable *self, PyObject *py_grad)
-       gradIsSparse = gradType == sparseType;
-  }
-
+  bool gradIsSparse = var.dtype() == grad.dtype() && toSparse(tensorTypeIdToBackend(var.type_id())) == tensorTypeIdToBackend(grad.type_id());
```
@li-roy this change is breaking backend extensions in that when the `var` backend type is a backend extension type, `toSparse()` just throws a runtime error. https://github.com/pytorch/pytorch/blob/master/c10/core/Backend.h#L31
